### ISSUE TITLE
Access violation during a memcpy

### ISSUE THREAD

**Issue by jflevesque-genetec** [openssl/openssl#18748]
Hi everyone,

We started having issues after we upgraded to 1.1.1 from 1.0.2p. We are currently on 1.1.1L, but we have seen the issue with an older version. The crash is always in the same method call, with pretty much the same inputs:
``` cpp
In rand_lib.c: 
/*
 * Add random bytes to the random pool.
 *
 * It is expected that the |buffer| contains |len| bytes of
 * random input which contains at least |entropy| bits of
 * randomness.
 *
 * Returns 1 if the added amount is adequate, otherwise 0
 */
int rand_pool_add(RAND_POOL *pool,
                  const unsigned char *buffer, size_t len, size_t entropy)
{
    if (len > pool->max_len - pool->len) {
        RANDerr(RAND_F_RAND_POOL_ADD, RAND_R_ENTROPY_INPUT_TOO_LONG);
        return 0;
    }

    if (pool->buffer == NULL) {
        RANDerr(RAND_F_RAND_POOL_ADD, ERR_R_INTERNAL_ERROR);
        return 0;
    }

    if (len > 0) {
        /*
         * This is to protect us from accidentally passing the buffer
         * returned from rand_pool_add_begin.
         * The check for alloc_len makes sure we do not compare the
         * address of the end of the allocated memory to something
         * different, since that comparison would have an
         * indeterminate result.
         */
        if (pool->alloc_len > pool->len && pool->buffer + pool->len == buffer) {
            RANDerr(RAND_F_RAND_POOL_ADD, ERR_R_INTERNAL_ERROR);
            return 0;
        }
        /*
         * We have that only for cases when a pool is used to collect
         * additional data.
         * For entropy data, as long as the allocation request stays within
         * the limits given by rand_pool_bytes_needed this rand_pool_grow
         * below is guaranteed to succeed, thus no allocation happens.
         */
        if (!rand_pool_grow(pool, len))
            return 0;
        memcpy(pool->buffer + pool->len, buffer, len);
        pool->len += len;
        pool->entropy += entropy;
    }

    return 1;
}
```
The memcpy crashes and the information I have in my crash dump has the buffer address being 0x00000010. Now, how this can be, I have no clue. In most crash dumps (but not all), the call originates from `rand_drbg_get_additional_data` which seems to only be called in `RAND_DRBG_bytes` (drbg_lib.c). From what I could see, the buffer address is supposed to represent an object created on the stack, so how it ends up with a 0x00000010 address value, I have no clue.
Note: another thing that makes no sense is that the count value passed to memcpy is 0 and `len` is already checked for greater than 0.

Since we cannot reproduce the issue and only happens with our customers, I was wondering what options we have to push the investigation forward? Is it only compilation flags that can help the debugging of this issue? 

Thanks in advance for anyone who contributes to this issue.

**Comment by paulidale** [comment]
I'm guessing pool->buffer is NULL.

The most likely reason is something is overwriting memory.

**Comment by slontis** [comment]
You could try logging all the info passed into the memcpy (just before the memcpy call) to see if everything looks ok.

**Comment by jflevesque-genetec** [comment]
Multithreaded application. From what I could determine, the application is not closing. This is the full call stack when it crashes:
![image](https://user-images.githubusercontent.com/39772820/178010877-73a8e4df-76cf-4c80-9c16-a6fd5a5503c5.png)
Not sure if its memory corruption, the dump file itself or code optimisation, but I don't have any locals to any stack frames except for the memcpy. And here's what I get:
![image](https://user-images.githubusercontent.com/39772820/178011105-825bff40-75ef-4c73-b82d-21f0a95606c9.png)
For reminder: `void* memcpy( void* dest, const void* src, std::size_t count );`
So:
- `src` == `buffer` which is supposed to be &data from the stack in `_rand_pool_add_additional_data`
- `count` == 0, which couldn't happen based on the code pasted in the description

At this point, are we all leaning for memory corruption or something else?

**Comment by mattcaswell** [comment]
Probably we need to see the value of `pool->buffer` and `pool->len` at the point that the `memcpy` crashes. The info you posted above doesn't show us that.

I agree with @paulidale that it seems likely that `pool->buffer` is NULL. However as we can see in the `rand_pool_add` code you posted there is already a check for this earlier in the function. So either (1) the assumption that `pool->buffer` is NULL is wrong and it instead contains some invalid value or (2) some other thread has made it NULL between the earlier check and the later use. Either way we could do with understanding the value so that we can figure out which of these 2 alternatives is happening.

**Comment by jflevesque-genetec** [comment]
I'll see what I can do to add logging in that method. Any compilation flags I should enable to help the investigation?

**Comment by kaduk** [comment]
`-ggdb3` might help let the debugger recover locals, and of course so would using a lower optimization level (but the latter might also chanbge behavior...)

**Comment by jflevesque** [comment]
We are not using gcc but we did have the equivalent already in our build pipeline. Seems like it may be memory corruption of some kind. I suggest closing the issue and if I have any other questions or updates, I'll try to reopen the issue.

**Comment by katvillariba** [comment]
Hello, 
We recently observed this exact crash on 1.1.1s with our multi-threaded application. @jflevesque did you figure out what was wrong?

pool->buffer is not NULL

```
(gdb) p *((RAND_POOL*)(0x7fffcc0c3b00))
$119 = {buffer = 0x7fffcc0c3180 "", len = 0, attached = 0, secure = 0, min_len = 0, max_len = 12288, alloc_len = 48, entropy = 0, entropy_requested = 0}
```

Here is the full backtrace:
```
#0  memcpy () at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:145
No locals.
#1  0x00007ffff6dede29 in rand_pool_add (pool=0x7fffcc0c3b00, buffer=0x7fff90fc7ef0 "G^", len=16, entropy=0) at ../../../../../crypto/rand/rand_lib.c:759
No locals.
#2  0x00007ffff6f0c52e in rand_pool_add_additional_data (pool=0x7fffcc0c3b00) at ../../../../../crypto/rand/rand_unix.c:796
        data = {fork_id = 24135, tid = 0, time = 2241750262}
#3  0x00007ffff6ded23a in rand_drbg_get_additional_data (pool=0x7fffcc0c3b00, pout=0x7fff90fc7f78) at ../../../../../crypto/rand/rand_lib.c:293
        ret = 0
#4  0x00007ffff6def57c in RAND_DRBG_bytes (drbg=0x7fffcc11cd50, out=0x7fffcc71b618 "", outlen=32) at ../../../../../crypto/rand/drbg_lib.c:661
        additional = 0x0
        additional_len = 140735625854896
        chunk = 140735625854880
        ret = 0
#5  0x00007ffff6defb78 in drbg_bytes (out=0x7fffcc71b618 "", count=32) at ../../../../../crypto/rand/drbg_lib.c:957
        ret = 32767
        drbg = 0x7fffcc11cd50
#6  0x00007ffff6dee1ed in RAND_bytes (buf=0x7fffcc71b618 "", num=32) at ../../../../../crypto/rand/rand_lib.c:945
        meth = 0x7ffff73853a0 <rand_meth>
#7  0x00007ffff6d98382 in ssl_fill_hello_random (s=0x7fffcc719cd0, server=0, result=0x7fffcc71b618 "", len=32, dgrd=DOWNGRADE_NONE) at ../../../../../ssl/s3_lib.c:4592
        send_time = 0
        ret = 32767
#8  0x00007ffff6ed0b86 in tls_construct_client_hello (s=0x7fffcc719cd0, pkt=0x7fff90fc8110) at ../../../../../ssl/statem/statem_clnt.c:1148
        p = 0x7fffcc71b618 ""
        sess_id_len = 140735625855152
        i = 1
        protverr = 0
        comp = 0x7fffcc719cd0
        sess = 0x0
        session_id = 0x7ffff6d9520d <ssl3_set_handshake_header+82> "\205?u\a?"
        __PRETTY_FUNCTION__ = "tls_construct_client_hello"
#9  0x00007ffff6ed8bf8 in write_state_machine (s=0x7fffcc719cd0) at ../../../../../ssl/statem/statem.c:843
        st = 0x7fffcc719d18
        ret = 32767
        transition = 0x7ffff6ecfc5e <ossl_statem_client_write_transition>
        pre_work = 0x7ffff6ed0007 <ossl_statem_client_pre_work>
        post_work = 0x7ffff6ed0137 <ossl_statem_client_post_work>
        get_construct_message_f = 0x7ffff6ed0513 <ossl_statem_client_construct_message>
        cb = 0x0
        confunc = 0x7ffff6ed0a1c <tls_construct_client_hello>
        mt = 1
        pkt = {buf = 0x7fffcca79170, staticbuf = 0x0, curr = 4, written = 4, maxsize = 18446744073709551615, subs = 0x7fffcc769da0}
#10 0x00007ffff6ed7fb6 in state_machine (s=0x7fffcc719cd0, server=0) at ../../../../../ssl/statem/statem.c:443
        buf = 0x0
        cb = 0x0
        st = 0x7fffcc719d18
        ret = -1
        ssret = 0
#11 0x00007ffff6ed7a7d in ossl_statem_connect (s=0x7fffcc719cd0) at ../../../../../ssl/statem/statem.c:250
No locals.
#12 0x00007ffff6d8a5c0 in SSL_do_handshake (s=0x7fffcc719cd0) at ../../../../../ssl/ssl_lib.c:3732
        ret = 1
#13 0x00007ffff6d07608 in _secure_finish_connect (hand=0x7fffcc71ba38) at socket_secure_intf.c:1561
        error = 268697739
        ret = 0
#14 0x00007ffff6d0882b in _secure_send (p_connection=0x7fffcc71ba38, buffer=0x7fffcc535e58, buffer_size=537, p_size_sent=0x7fff90fc82c8) at socket_secure_intf.c:2110
        hand = 0x7fffcc71ba38
        error = 0
        sent = 0
        sslerr = 0
#15 0x00007ffff6d092e3 in _secure_iostream_write_buffer (callback_data=0x7fffcc71ba38, buffer=0x7fffcc535e58, buffer_size=537, p_bytes_written=0x7fff90fc8318) at socket_secure_intf.c:2610
        p_connection = 0x7fffcc71ba38
        bytes_written = 0
        error = 0
#16 0x00007ffff6f7aba5 in gcsl_iostream_write (h_iostream=0x7fffcc95afb8, data_buffer=0x7fffcc535e58, buffer_size=537, p_bytes_written=0x7fff90fc83c0) at gcsl_iostream.c:659
        p_iostream = 0x7fffcc95afb8
        bytes_written = 0
        error = 32767
#17 0x00007ffff6cffd56 in gcsl_socket_iostream_write (h_iostream=0x7fffcc95afb8, send_buffer=0x7fffcc535e58, send_size=537, size_sent=0x7fff90fc83c0, timeout_ms=30000) at gcsl_socket_iostream.c:168
        error = 0
        elapsed_ms = 0
        iostream_handle = 0x7fffcc95afb8
#18 0x00007ffff6c91fa7 in _http_send (p_http_conn=0x7fffcc535be8, buffer=0x7fffcc535e58, send_size=537, p_size_sent=0x7fff90fc8420) at gcsl_http_impl.c:645
        size_sent = 0
        buffer_bytes = 0x7fffcc535e58 "POST /gcsp/2.0 HTTP/1.1\r\nHost: c7991552.mgr-mid.gcsp.cddbp.net:443\r\nConnection: close\r\nContent-Length: 777\r\nUser-Agent: GCSL_GCSP 3.18.0\r\nContent-Type: application/x-gn-message\r\nX-GN-Protocol: gcsp; v"...
        socket_state = 33
        total_sent = 0
        error = 0
#19 0x00007ffff6c91abd in http_impl_send_content (p_connection_data=0x7fffcc535be8, content_buffer=0x7fffcc11e1a8, content_length=777) at gcsl_http_impl.c:472
        p_http_conn = 0x7fffcc535be8
        byte_len = 537
        size_sent = 140735625856144
        sz_request = 0x7fffcc535e58 "POST /gcsp/2.0 HTTP/1.1\r\nHost: c7991552.mgr-mid.gcsp.cddbp.net:443\r\nConnection: close\r\nContent-Length: 777\r\nUser-Agent: GCSL_GCSP 3.18.0\r\nContent-Type: application/x-gn-message\r\nX-GN-Protocol: gcsp; v"...
        error = 0
#20 0x00007ffff6c8e59f in _http_request_send_content (p_http_conn=0x7fffcc34ba98, content_buffer=0x7fffcc11e1a8, content_length=777) at gcsl_http.c:2241
        content_byte_buffer = 0x7fffcc11e1a8 "XGNC00244201000000320010000002930010\224"
        content_buffer_pos = 0
        b_abort = 0 '\000'
        error = 0
#21 0x00007ffff6c8bb1b in gcsl_http_send_request (connection=0x7fffcc34ba98, send_method=0x7ffff6fba697 "POST", target_url=0x7fffcc2503f8 "https://c7991552.mgr-mid.gcsp.cddbp.net:443/gcsp/2.0", header_values=0x7fffcc59d2e8, content_buffer=0x7fffcc11e1a8, content_length=777) at gcsl_http.c:1143
        p_http_conn = 0x7fffcc34ba98
        accum_headers = 0x0
        target_path = 0x7fffcc250423 "/gcsp/2.0"
        port = 443
        b_prefixed = 1 '\001'
        headers_size = 513
        headers = 0x7fffcc130308 ""
        hostname = "c7991552.mgr-mid.gcsp.cddbp.net", '\000' <repeats 224 times>
        accum_buf = "\b\003\023??\177\000\000\000\002\000\000\000\000\000\000?\003\000\000\000\000\000\000\000\001\000\000\000\000\000\000Host: c7991552.mgr-mid.gcsp.cddbp.net:443\r\nConnection: close\r\nContent-Length: 777\r\nUser-Agent: GCSL_GCSP 3.18.0\r\nContent-Type: application/x-gn-message\r\nX-GN-Protocol: "...
        b_abort = 0 '\000'
        b_https = 1 '\001'
        error_local = 949
        error = 0
#22 0x00007ffff6c3ca3a in gcsp_connection_send (p_connection=0x7fffcc59d198, request_flags=3, target_path=0x0, content_buffer=0x7fffcc11e1a8, content_length=777, response_callback=0x7ffff6c422cc <_gcsp_transaction_response_callback>, callback_data=0x7fffcc59d0a8) at gcsp_connection.c:529
        p_response_handler = 0x7fffcc1210b8
        p_mime_keys = 0x0
        host_url = 0x7fffcc2503f8 "https://c7991552.mgr-mid.gcsp.cddbp.net:443/gcsp/2.0"
        override_url = 0x0
        host_name = "p\210?\220?\177\000\000\230\023S??\177\000\000\003\000\000\000\000\000\000\000?\001%??\177\000\000\060\210?\220?\177\000\000\230AY?\000\000\000\000 \210?\220?\177\000\000?.???\177\000\000\000\000\000\000\000\000\000\000\230\023S??\177\000\000\220\210?\220?\177\000\000K????\177\000\000\220\210?\220?\177\000\000?\210?\220?\177\000\000?\210?\220?\177\000\000\001???\001\000\000\000\213????\177\000\000\210@Y??\177\000\000\220\210?\220", '\000' <repeats 12 times>, "HeY??\177\000\000\216\006\000\000\000\000\000\000\230\023S??\177\000\000\000K7\a\r>b8?\210?\220?\177\000\000"...
        host_port = 52307
        b_clb = 0 '\000'
        b_compression = 1 '\001'
        b_encryption = 1 '\001'
        b_https = 1 '\001'
        content_buffer_del = 0x7fffcc11e1a8 "XGNC00244201000000320010000002930010\224"
        content_buffer_tmp = 0x7fffcc11e1a8 "XGNC00244201000000320010000002930010\224"
        content_length_tmp = 777
        value = 0x7fffcc5914e8 "gcsp_encryption_algorithm_tls"
        error = 0
#23 0x00007ffff6c41f72 in _gcsp_execute_transaction (p_transaction=0x7fffcc59d0a8) at gcsl_gcsp_transaction.c:1382
        requests_xml_str = 0x7fffcc596548 "<?xml version=\"1.0\" encoding=\"UTF-8\" ?>\n<REQUESTS>\n  <REQUEST CMD=\"HELLO\" OUTPUT=\"SESSION_ID\" IDENT=\"HELLO\">\n    <CLIENT>\n      <IDENT>7991552</IDENT>\n      <IDENT_TAG>48A302BEC7395871114AC11C13933FB8"...
        in_payload = 0x7fffcc596548 "<?xml version=\"1.0\" encoding=\"UTF-8\" ?>\n<REQUESTS>\n  <REQUEST CMD=\"HELLO\" OUTPUT=\"SESSION_ID\" IDENT=\"HELLO\">\n    <CLIENT>\n      <IDENT>7991552</IDENT>\n      <IDENT_TAG>48A302BEC7395871114AC11C13933FB8"...
        in_payload_size = 1678
        b_retry = 0 '\000'
        error = 0
        cancelled = 0
        timing = 0
#24 0x00007ffff6c3f371 in gcsl_gcsp_transaction_execute (transaction_handle=0x7fffcc59d0a8) at gcsl_gcsp_transaction.c:207
        p_transaction = 0x7fffcc59d0a8
        error = 0
#25 0x00007ffff6c190ee in _sdkmgr_lookup_gcsp_execute_gcsp_requests (p_lookup=0x7fffcc17f828) at sdkmgr_impl_lookup_gcsp.c:2870
        error = 0
        storage_error = 0
#26 0x00007ffff6c1a8ff in _sdkmgr_lookup_gcsp_execute (p_lookup_data=0x7fffcc17f828, user_handle=0x55555588e838, locale_handle=0x555555891ac8) at sdkmgr_impl_lookup_gcsp.c:3405
        p_lookup = 0x7fffcc17f828
        p_lookup_request = 0x7fffcc3226c8
        p_followup_lookup = 0x0
        lookup_request_count = 1
        req = 1
        b_autoreg = 0 '\000'
        p_void = 0x7fffcc3226c8
        client_id = 0x55555588f168 "7991552"
        client_id_tag = 0x55555588f2a8 "48A302BEC7395871114AC11C13933FB8"
        user_value = 0x55555588ec48 "648065833567575395"
        option_value = 0x0
        permission_error = 2424308769
        error = 0
#27 0x00007ffff6bda30a in _sdkmgr_lookup_execute (lookup_handle=0x7fffcc528518, user_handle=0x55555588e838, locale_handle=0x555555891ac8) at sdkmgr_intf_lookup.c:735
        p_lookup = 0x7fffcc528518
        error = 0
#28 0x00007ffff7ad5267 in _musicid_request_perform_text (p_musicid=0x7fffcc51d3a8, lookup_type=0x7ffff7b2bbae "gnsdk_lookup_type_text_match", p_response_album_gdo=0x7fff90fc8b90) at gnsdk_musicid_requests.c:215
        error = 0
        lookup_interface = 0x7fffcc17f6c8
        lookup_handle = 0x7fffcc528518
        response_album_gdo = 0x0
        permission_status = 0
#29 0x00007ffff7ac61ac in _musicid_query_find_matches (musicid_query_handle=0x7fffcc51d3a8, p_response_gdo=0x7fff90fc8c60) at mid_api_impl.c:2205
        stats_handle = 0x0
        p_musicid = 0x7fffcc51d3a8
        p_musicid_gdo_response = 0x0
        response_gdo = 0x0
        pref_lang = 0x0
        lookup_value_map = 0x7fffcc186018
        error = 0
        error_api = 1431792032
        stats_error = 0
#30 0x00007ffff7ae223e in gnsdk_musicid_query_find_matches (musicid_query_handle=0x7fffcc51d3a8, p_response_gdo=0x7fff90fc8c60) at gnsdk_musicid.c:915
        error = 32767
#31 0x0000555555568722 in gracenote::musicid::GnMusicId::FindMatches (this=0x7fff90fc8d70, albumTitle=0x555555574f64 "All Eyez On Me [Disc 1]", trackTitle=0x555555574f4a "2 Of Amerikaz Most Wanted", albumArtistName=0x555555574f45 "2Pac", trackArtistName=0x0, composerName=0x0) at ../../src_wrapper/gnsdk_musicid.cpp:743
        response_gdo = 0x0
        error = 21845
#32 0x000055555555d13c in doSampleTextQuery (user=..., lookupMode=gracenote::kLookupModeOnlineNoCache, album_title=0x555555574f64 "All Eyez On Me [Disc 1]", track_title=0x555555574f4a "2 Of Amerikaz Most Wanted", artist_name=0x555555574f45 "2Pac") at main.cpp:265
        musicid = {<gracenote::GnObject<gnsdk_musicid_query_handle_t_s*>> = {_vptr.GnObject = 0x555555782eb0 <vtable for gracenote::musicid::GnMusicId+16>, handle_ = 0x7fffcc51d3a8}, eventHandler_ = {_vptr.GnManagedPtr = 0x555555782f68 <vtable for gracenote::GnManagedPtr<gracenote::IGnStatusEvents*>+16>, handle_ = 0x0, scheme_ = gracenote::GnManagedPtr<gracenote::IGnStatusEvents*>::ptr_nomanage}, options_ = {weakhandle_ = 0x7fffcc51d3a8}, cancelled_ = false}
        match_response = {<gracenote::metadata::GnDataObject> = {<gracenote::GnObject<gnsdk_gdo_handle_t_s*>> = {_vptr.GnObject = 0x0, handle_ = 0x0}, <No data fields>}, <No data fields>}
        count = 0
#33 0x000055555555d7df in <lambda()>::operator()(void) const (__closure=0x555556109898) at main.cpp:467
        i = 77
        user = @0x7fffffffd6b0: {<gracenote::GnObject<gnsdk_user_handle_t_s*>> = {_vptr.GnObject = 0x555555783060 <vtable for gracenote::GnUser+16>, handle_ = 0x55555588e838}, options_ = {_online_enc_options = {weakhandle_ = 0x0}, weakhandle_ = 0x55555588e838}, shared_ = {_vptr.GnManagedPtr = 0x555555783098 <vtable for gracenote::GnManagedPtr<gracenote::GnUser::Shared*>+16>, handle_ = 0x555555852b70, scheme_ = gracenote::GnManagedPtr<gracenote::GnUser::Shared*>::ptr_manage}}
        lookupMode = @0x7fffffffd624: gracenote::kLookupModeOnline
#34 0x000055555555fd8d in std::__invoke_impl<void, main(int, char**)::<lambda()> >(std::__invoke_other, <lambda()> &&) (__f=...) at /usr/include/c++/7/bits/invoke.h:60
No locals.
#35 0x000055555555fa08 in std::__invoke<main(int, char**)::<lambda()> >(<lambda()> &&) (__fn=...) at /usr/include/c++/7/bits/invoke.h:95
No locals.
#36 0x0000555555560872 in std::thread::_Invoker<std::tuple<main(int, char**)::<lambda()> > >::_M_invoke<0>(std::_Index_tuple<0>) (this=0x555556109898) at /usr/include/c++/7/thread:234
No locals.
#37 0x00005555555607b4 in std::thread::_Invoker<std::tuple<main(int, char**)::<lambda()> > >::operator()(void) (this=0x555556109898) at /usr/include/c++/7/thread:243
No locals.
#38 0x0000555555560534 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<main(int, char**)::<lambda()> > > >::_M_run(void) (this=0x555556109890) at /usr/include/c++/7/thread:186
No locals.
#39 0x00007ffff67f36df in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.
#40 0x00007ffff651e6db in start_thread (arg=0x7fff90fc9700) at pthread_create.c:463
        pd = 0x7fff90fc9700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140735625860864, 7181917144278407410, 140735625859008, 0, 93825004509328, 140737488344176, -7181998483979593486, -7181931661514991374}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
#41 0x00007ffff5a8961f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

**Comment by katvillariba** [comment]
@jflevesque-genetec Thank you so much for your insight. You have saved us from a lot of headaches. We took your advice and looked into our OpenSSL configuration. It appears that there is a bug in the OpenSSL Configure file, such that specifying the --static option was causing the --nothreads option to be applied even though we specified --threads. We have not observed the crash since we modified the OpenSSL Configure file by applying the workaround described in #14574 .
